### PR TITLE
Issue 868 archive data

### LINKED ISSUE #868: Archive studygroups and discussions on `delete my account`.
r/n when user deletes their account we are archiving their [Hangouts](https://github.com/codebuddies/codebuddies/blob/3335e76cbbc325175abdd68f97e09d23399b412a/server/archived_users/methods.js#L49), [Learnings](https://github.com/codebuddies/codebuddies/blob/3335e76cbbc325175abdd68f97e09d23399b412a/server/archived_users/methods.js#L50) and other informations related to that user.

Similarly archive their Discussions and StudyGroups.

### PR DESCRIPTION
**Pull request by railsstudent**
Fixes #868,

Remove study groups, discussions and discussion responses when account is deleted.
Set visibility of discussion to false  when author.id is user id.
Set visibility of discussion response to false when author.id is user id
Set visibility of study group to false when members.$.role is owner and members.$.id is user id.

@distalx, authRevoke function revokes authorization of slack app only. Should we revoke authorization of github app too? Thanks.
 
```
authRevoke = function(accessToken) {
  check(accessToken, String);

  var response = HTTP.get("https://slack.com/api/auth.revoke",
    {params:
      {token: accessToken,
       test: false
      }
    });
  return response.data.ok && response.data;
};
```

### CODE REVIEW

**Comment by sergeant-q** [comment]
As a side note, feel free to squash your commits when you're ready to merge in your changes if it makes sense.

**Comment by lpatmo** [comment]
Why are you so amazing, @railsstudent 

If I sign in using Github and try to archive my account, I get a 500 error. Not sure if it's related to the revocation of permissions that you were talking about.

@sergeant-q Noted. I've usually chosen `squash and merge` for the merge option because I like the readability.

**Comment by railsstudent** [comment]
@lpatmo  I did not test with github account,  I will take a look. Thanks.

**Comment by railsstudent** [comment]
@lpatmo  Is there a git command to do the squash and merge? I use gitkraken gui client to merge code from upstream to my local repo and then push my code changes to remote repo.

**Comment by sergeant-q** [comment]
@railsstudent when you're ready to merge your PR, you can either do a straight forward merge or select from the merge button drop down to "merge and squash"

However, this will squash all your commits from your feature branch into one, thus losing all your commit history which I'm personally against. 

You can however, squash your commits on your feature branch using git rebase or reset before opening up your PR to clean things up.

**Comment by lpatmo** [comment]
@sergeant-q You meant that advice for the PR reviewer, right? Since the person who made the PR shouldn't generally approve and merge in their own PR.

(I've admittedly broken that rule for a couple of my own PRs, so I'll try to stop.)

**Comment by lpatmo** [comment]
Slightly off-topic: TIL about https://stackoverflow.com/questions/5189560/squash-my-last-x-commits-together-using-git from @sergeant-q. That's what he meant by "squashing your commits using git rebase or git reset." I didn't know before that `git rebase -i [hash]` could squash commits.

**Comment by wuworkshop** [comment]
@railsstudent Slightly off topic, but just in case you weren't aware, I came across this page and video from the GitKraken documentation about Squash:

https://support.gitkraken.com/working-with-commits/squash

@lpatmo From that SO thread you linked, would another CLI command be `git merge --squash`?

Possible reference: https://git-scm.com/docs/merge-options#merge-options---squash

**Comment by stain88** [comment]
`git merge --squash` is just that, a merge with the `squash` flag.
`git rebase -i` is a more interactive tool, one use of which is to (potentially) squash all your commits _before_ the merge, but also has options to cherry-pick certain commits, ignore others, etc...

**Comment by railsstudent** [comment]
@sergeant-q @lpatmo  Completed the PR and github oauth app can be revoked successfully.
However, I cannot git squash all the "Merge remote-tracking branch 'upstream/staging' into staging" commits in the feature branch

This is the output of git merge --squash feature/issue-868-archive-data in my terminal.
merge: feature/issue-868-archive-data - not something we can merge

**Comment by sergeant-q** [comment]
@railsstudent are you able to squash your commits on your own branch? That's what you're aiming for.

Once you do and clean up your branch you can repush and this PR will be updated. Then we can review and approve. Hope that makes sense. 

As a side note. I'm not sure if you tested this but when I click on "delete my account" and the ouath revoke goes through successfully, the app doesn't sign me out. I'm still logged into CB. 

Shouldn't I not have an account at this point? Not sure if this should be addressed as a separate issue or fix it here. Thoughts?

**Comment by railsstudent** [comment]
@sergeant-q  I tested it and when I clicked anywhere else in Codebuddies, I was logged out.

File: server/archived_users/methods.js [hunk 1]
```diff
@@ -1,69 +1,87 @@
...
+      );
     }
 
-    //removing access for codebuddies slack app
-    const result = authRevoke(accessToken);
-    if(result.ok){
+    const user = Meteor.users.findOne({ _id: userId });
+    let accessToken = null;
+    let result = null;
+    if (user.services.slack && user.services.slack.accessToken) {
```
Comment (distalx) [inline server/archived_users/methods.js]: When user has used both services for sign in (slack & github), user would have both of them available. In that case  system will only revoke the slack token and not revoke the github one. 
Here, we should check for both services separately and revoke them accordingly.  

Very roughly :
```
if (slackTokenExists) {
 // remove slack auth
}
if (githubTokenExists) {
 // remove github auth
}
```

**Comment by lpatmo** [comment]
@sergeant-q Out of curiosity, is `squashing commits locally` something you want to promote for all future PRs? I haven't done it before in the past, and only learned that it was a thing someone could do from this PR thread. 

Will look into the docker issue.

@distalx Thanks for reviewing!

File: server/archived_users/methods.js (hunk 1 above)
Comment (connieforeverdev) [reply server/archived_users/methods.js]: Will fix that later

Comment (connieforeverdev) [reply server/archived_users/methods.js]: Will fix that later

Comment (railsstudent) [reply server/archived_users/methods.js]: Will fix it later.

**Comment by railsstudent** [comment]
@sergeant-q @lpatmo  There are over 30 "Merge remote-tracking branch 'upstream/staging' into staging" commits in my staging branch that I do not how to erase.

This is my practice:
Before tackling any issue in the project, I merge upstream/staging to my local staging branch and push to origin/staging remote repo. The action always results to a "upstream/staging..." commit.   Next, I create a branch from staging, start working and push changes to remote repo.  Finally, I make pull request against codebuddies repo and boom 30+ commits listed in the PR.

Should I clone my github repo into a separate local directory and start anew? Thanks.

**Comment by lpatmo** [comment]
@sergeant-q Got it, thanks for explaining!

I've had the same problem from a repo that's cloned directly from the main repo (not a fork!). On a recent PR I submitted (https://github.com/codebuddies/codebuddies/pull/905), it looks like there were a bunch of unrelated commits on it, but the actual files changed are only https://github.com/codebuddies/codebuddies/pull/905/files. Same for other PRs I've submitted like https://github.com/codebuddies/codebuddies/pull/708.

As a reviewer, I typically don't look at individual commits because I know new commits can sometimes override old commits; I only look at the "Files Changed" tab to see the overall `diff` that the PR is introducing. So I suppose whether or not one is bothered by the mysterious extra commits depends on one's review style. :) 

Not sure how to override the extra `Merge remote-tracking branch 'upstream/staging' into staging` that Connie and I both get every time; it probably is related to how we both `git merge upstream/staging`, maybe? I glanced through some past PRs and noticed that some have them, and some don't. Given the ability to see the overall diff of each PR, I'm not too worried about it, though.

**Comment by lpatmo** [comment]
Super. Thanks so much, @railsstudent!!

**Pull request merged**